java-checks/src/main/java/org/sonar/java/checks/spring/SpringConfigurationWithAutowiredFieldsCheck.java

```diff
@@ -0,0 +1,117 @@
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3305")
+public class SpringConfigurationWithAutowiredFieldsCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE_FORMAT = "Inject this field value directly into \"%s\", the only method that uses it.";
+
+  private static final String CONFIGURATION_ANNOTATION = "org.springframework.context.annotation.Configuration";
+  private static final String BEAN_ANNOTATION = "org.springframework.context.annotation.Bean";
+  private static final List<String> AUTOWIRED_ANNOTATIONS = Arrays.asList(
+    "org.springframework.beans.factory.annotation.Autowired",
+    "javax.inject.Inject");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    if (hasSemantic() && classTree.symbol().metadata().isAnnotatedWith(CONFIGURATION_ANNOTATION)) {
+      AutowiredFieldVisitor autowiredFieldVisitor = new AutowiredFieldVisitor();
+      classTree.accept(autowiredFieldVisitor);
+      BeanMethodVisitor beanMethodVisitor = new BeanMethodVisitor(autowiredFieldVisitor.autowiredFields.keySet());
+      classTree.accept(beanMethodVisitor);
+      beanMethodVisitor.beanMethodsThatUseAutowiredFields.entrySet().stream()
+        .filter(beanMethodsThatUseAutowiredField -> beanMethodsThatUseAutowiredField.getValue().size() == 1)
+        .forEach(beanMethodsThatUseAutowiredField -> reportIssue(
+          autowiredFieldVisitor.autowiredFields.get(beanMethodsThatUseAutowiredField.getKey()).simpleName(),
+          String.format(MESSAGE_FORMAT, beanMethodsThatUseAutowiredField.getValue().get(0))));
+    }
+  }
+
+  private static class AutowiredFieldVisitor extends BaseTreeVisitor {
+    private final Map<Symbol, VariableTree> autowiredFields = new HashMap<>();
+
+    @Override
+    public void visitVariable(VariableTree tree) {
+      Symbol symbol = tree.symbol();
+      if (symbol.owner().isTypeSymbol() && AUTOWIRED_ANNOTATIONS.stream().anyMatch(a -> symbol.metadata().isAnnotatedWith(a))) {
+        autowiredFields.put(symbol, tree);
+      }
+    }
+  }
+
+  private static class BeanMethodVisitor extends BaseTreeVisitor {
+    private final Map<Symbol, List<String>> beanMethodsThatUseAutowiredFields = new HashMap<>();
+
+    BeanMethodVisitor(Set<Symbol> autowiredFields) {
+      autowiredFields.forEach(f -> beanMethodsThatUseAutowiredFields.put(f, new ArrayList<>()));
+    }
+
+    @Override
+    public void visitMethod(MethodTree methodTree) {
+      if (methodTree.symbol().metadata().isAnnotatedWith(BEAN_ANNOTATION)) {
```

Comment:
I think we should also visit methods without annotation and do some logic to avoid the following FP:
```
@Configuration
class A {

  @Autowired private Bar multipleUsage; // false-positive

  @Bean
  public Foo method2() {
    return extractedMethod();
  }

  private Foo extractedMethod() {
    return new Foo(this.multipleUsage);
  }

  @Bean
  public Foo method3() {
    return new Foo(this.multipleUsage);
  }

}
```